### ISSUE TITLE
[Feature] Enclosures as part of greader API

### ISSUE THREAD

**Issue by jangernert** [FreshRSS/FreshRSS#4702]
**Is your feature request related to a problem? Please describe.**
It would be nice to easily access enclosures like thumbnails, podcast audio tracks etc. from client applications without parsing the html content.

**Describe the solution you’d like**
Extend the google reader API entries with enclosures like most other APIs do. Examples would be the [feedly API](https://developer.feedly.com/v3/entries/), [Nextcloud News API](https://github.com/nextcloud/news/blob/master/docs/api/api-v1-3.md#items), [tt-rss API (called attachments)](https://tt-rss.org/wiki/ApiReference) and [feedbin API](https://github.com/feedbin/feedbin-api/blob/master/content/entries.md).

**Describe alternatives you’ve considered**
It seems like enclosures are somehow embedded in the article html content:
```
<div class=\"enclosure\"><p class=\"enclosure-title\">💧🔥 BEST COMBO COMMANDER | Jhoira, Ageless Innovator | Competitive Historic Brawl [MTG Arena]</p><p><img class=\"enclosure-thumbnail\" src=\"https://i4.ytimg.com/vi/_qUe1nd09Ic/hqdefault.jpg\" alt=\"\" /></p><p class=\"enclosure-content\">
```

So in theory one could parse the content and extract the data from there. But that would mean a lot of special handling for FreshRSS compared to other popular feed services in the client app.

**Additional context**
Issue filed against NewsFlash client: https://gitlab.com/news-flash/news_flash_gtk/-/issues/374

The google reader API has already been extended by others with minor new features (e.g. friends lists https://github.com/theoldreader/api#friends).

**Comment by Alkarex** [comment]
The Google Reader API already has an `enclosure` section.

Example from FreshRSS:

```json
{
  "id": "tag:google.com,2005:reader/item/0005deaa12dfd1fc",
  "crawlTimeMsec": "1652196926018",
  "timestampUsec": "1652196926018044",
  "published": 1652185080,
  "title": "The new Portenta X8 (with Linux!) and Max Carrier redefine what’s possible",
  "summary": {
    "content": "<figure data-sanitized-class=\"wp-block-image size-large\"><div data-sanitized-class=\"image-post\"><img loading=\"lazy\" width=\"1024\" height=\"549\" src=\"https://blog.arduino.cc/wp-content/uploads/2022/05/X8-Blogspot-cover-1024x549.png\" alt=\"Portenta X8 Max Carrier\" data-sanitized-class=\"wp-image-31357\"></div></figure>\n\n\n\n<p>Arduino’s mission has always been to give creators, makers and innovators the tools they need to turn their ideas into real projects. That will never change. What <em>is</em> changing, and at warp speed, is our definition of possible.</p>\n\n\n\n<p>Because every day our accessible, flexible and reliable open-source hardware grows more powerful, and our ecosystem more complete. As you may have heard during Arduino Week, Arduino Pro is launching two new products in the Portenta range: the <a href=\"https://www.arduino.cc/pro/hardware/product/portenta-x8\" target=\"_blank\" rel=\"noreferrer noopener\">revolutionary X8 module</a>, which <strong>merges Arduino and Linux for the first time</strong>, and <a href=\"https://www.arduino.cc/pro/hardware/product/portenta-max-carrier\" target=\"_blank\" rel=\"noreferrer noopener\">Max Carrier</a>, which gives you super prototyping powers to <strong>make your ideas come to life easier and faster than ever</strong>. </p>\n\n\n\n<figure data-sanitized-class=\"wp-block-embed is-type-video is-provider-youtube wp-block-embed-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio\"><div data-sanitized-class=\"wp-block-embed__wrapper\">\n<iframe loading=\"lazy\" title=\"Introducing the Portenta X8 by Arduino Pro\" width=\"500\" height=\"281\" src=\"https://www.youtube.com/embed/c2nz1wLIp5g?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen sandbox=\"allow-scripts allow-same-origin\"></iframe>\n</div></figure>\n\n\n\n<p>Portenta X8 is a plug-and-play, <strong>industrial-grade</strong> SOM that comes with <strong>Linux OS preloaded</strong> onboard, making for a hybrid combination of microprocessor and microcontroller with the capability for AI and ML on the edge. It’s basically <strong>two products in one</strong>, with the power of no less than <strong>9 cores</strong>. It features an NXP<sup>®</sup> i.MX 8M Mini Cortex<sup>®</sup>-A53 quad-core, up to 1.8GHz per core + 1x Cortex<sup>®</sup>-M4 up to 400MHz, plus the STM32H747XI dual-core Cortex<sup>®</sup>-M7 up to 480Mhz +M4 32 bit Arm<sup>®</sup> MCU up to 240Mhz, and is capable of running device-independent software thanks to its modular container architecture.</p>\n\n\n\n<p>With <strong>onboard WiFi/Bluetooth<sup>®</sup></strong> connectivity you can carry out <strong>OS/application updates remotely</strong>, so the Linux kernel environment is always at top performance levels. Enhanced <strong>security</strong> is guaranteed by X8’s NXP SE050C2 Crypto element – keeping connections secure at the hardware level with <strong>PSA certification</strong>. The module has also achieved <strong>Arm SystemReady certification</strong> and integrated Parsec services, making it one of the first Cassini Products available to you on the market.</p>\n\n\n\n<figure data-sanitized-class=\"wp-block-video\"><video controls=\"controls\" loop src=\"https://www.datocms-assets.com/19535/1647956353-x8-chips-explode-for-website.mp4\" preload=\"none\"></video></figure>\n\n\n\n<p>It’s everything you need to develop your most ambitious projects for Industry 4.0, smart agriculture, connected buildings and smart cities: check out the <a href=\"https://www.arduino.cc/pro/hardware/product/portenta-x8\" target=\"_blank\" rel=\"noreferrer noopener\">full technical specs on the dedicated page</a> and <strong>superpower your Linux applications with real-time execution.</strong></p>\n\n\n\n<p>If that wasn’t enough, we are also introducing Portenta Max Carrier – which boosts X8 or H7 by adding connectivity options (Fieldbus, LoRa<sup>®</sup>,  Cat-M1 and NB-IoT), industrial connectors such as <strong>RS232/422/485, USB, mPCIe</strong> – plus three integrated <strong>audio jacks</strong>, a <strong>MicroSD</strong> card and more. </p>\n\n\n\n<p>You can combine this Arduino Pro carrier with existing Portenta modules to turn them into single-board computers or reference designs for Industry 4.0 – <strong>bringing your deployment time virtually down to zero</strong>. This allows you to swiftly prototype and develop high-performance projects such as remote control of industrial machinery and equipment, smart digital kiosks that enhance users’ experience, and custom HMI dashboards to control the smart appliances, lights and systems in your home or office remotely.</p>\n\n\n\n<figure data-sanitized-class=\"wp-block-embed is-type-video is-provider-youtube wp-block-embed-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio\"><div data-sanitized-class=\"wp-block-embed__wrapper\">\n<iframe loading=\"lazy\" title=\"Portenta Max Carrier from Arduino Pro\" width=\"500\" height=\"281\" src=\"https://www.youtube.com/embed/GxN2-hs9288?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen sandbox=\"allow-scripts allow-same-origin\"></iframe>\n</div></figure>\n\n\n\n<p>As excited as we are about these launches, it’s never just about a new product (or two!). <strong>It’s about how many new ideas we can spark, the innovation we can fuel, the new opportunities you can create, if you have the right tools.</strong></p>\n\n\n\n<p><a href=\"https://store.arduino.cc/products/portenta-x8\">Portenta X8</a> is available on the Store and <a href=\"https://store.arduino.cc/products/portenta-max-carrier\" target=\"_blank\" rel=\"noreferrer noopener\">Max Carrier</a> will arrive soon. We can’t wait to see the great innovations you will bring to life.</p>\n\n\n\n<p>Want to know more? <a href=\"https://www.arduino.cc/pro/hardware/product/portenta-max-carrier\">Read all the details</a> and judge for yourself.</p>\n<p>The post <a rel=\"nofollow\" href=\"https://blog.arduino.cc/2022/05/10/the-new-portenta-x8-with-linux-and-max-carrier-redefine-whats-possible/\">The new Portenta X8 (with Linux!) and Max Carrier redefine what’s possible</a> appeared first on <a rel=\"nofollow\" href=\"https://blog.arduino.cc/\">Arduino Blog</a>.</p><div class=\"enclosure\"><p class=\"enclosure-content\"><video preload=\"none\" src=\"https://www.datocms-assets.com/19535/1647956353-x8-chips-explode-for-website.mp4\" data-length=\"13766774\" data-type=\"video/mp4\" controls=\"controls\"></video> <a download=\"\" href=\"https://www.datocms-assets.com/19535/1647956353-x8-chips-explode-for-website.mp4\">💾</a></p></div>\n"
  },
  "canonical": [
    {
      "href": "https://blog.arduino.cc/2022/05/10/the-new-portenta-x8-with-linux-and-max-carrier-redefine-whats-possible/"
    }
  ],
  "alternate": [
    {
      "href": "https://blog.arduino.cc/2022/05/10/the-new-portenta-x8-with-linux-and-max-carrier-redefine-whats-possible/"
    }
  ],
  "categories": [
    "user/-/state/com.google/reading-list",
    "user/-/label/Blogs électronique",
    "user/-/label/Featured",
    "user/-/label/Max Carrier",
    "user/-/label/Portenta",
    "user/-/label/X8",
    "user/-/label/Portenta X8"
  ],
  "origin": {
    "streamId": "feed/173",
    "htmlUrl": "https://blog.arduino.cc/",
    "title": "Arduino Blog"
  },
  "enclosure": [
    {
      "href": "https://www.datocms-assets.com/19535/1647956353-x8-chips-explode-for-website.mp4",
      "type": "video/mp4",
      "length": 13766774
    }
  ],
  "author": "Arduino Team"
}
```

Example of query to find examples from your own FreshRSS instance:

```sh
 curl -s -H 'Authorization:GoogleLogin auth=alice/???' 'https://freshrss.example.net/api/greader.php/reader/api/0/stream/contents/user%2F-%2Fstate%2Fcom.google%2Freading-list?n=10000&r=n' | jq '.items[] | select(.enclosure) | .'
```

**Comment by Alkarex** [comment]
P.S. At the moment, it is only repeating the enclosures provided by the upstream feed, of which type is not image

**Comment by Alkarex** [comment]
Ref https://github.com/FreshRSS/FreshRSS/pull/2898
The current implementation is thought mostly for clients supporting podcasts. Clients supporting thumbnails are typically picking the first image in the HTML content. But we could / should improve our implementation of enclosures

**Comment by jangernert** [comment]
@Alkarex thanks for the info

Where does FreshRSS get the youtube thumbnail from? Because when I parse a youtube feed with [`feed-rs`](https://github.com/feed-rs/feed-rs) it does include thumbnails as enclosure/media

```
Entry {
    id: "yt:video:EwEUvsOnKBI",
    title: Some(
        Text {
            content_type: "text/plain",
            src: None,
            content: "WORST BIRTHDAY / Laelia / Zedruu / Xyris / Isshin / The Worst Possible Commander Show 23 Magic MTG",
        },
    ),
    updated: Some(
        2022-08-23T07:46:36Z,
    ),
    authors: [
        Person {
            name: "CovertGoCrew",
            uri: Some(
                "https://www.youtube.com/channel/UCuvVwV414lLmhgWzWpJbmGg",
            ),
            email: None,
        },
    ],
    content: None,
    links: [
        Link {
            href: "https://www.youtube.com/watch?v=EwEUvsOnKBI",
            rel: Some(
                "alternate",
            ),
            media_type: None,
            href_lang: None,
            title: None,
            length: None,
        },
    ],
    summary: None,
    categories: [],
    contributors: [],
    published: Some(
        2022-08-14T17:00:09Z,
    ),
    source: None,
    rights: None,
    media: [
        MediaObject {
            title: Some(
                Text {
                    content_type: "text/plain",
                    src: None,
                    content: "WORST BIRTHDAY / Laelia / Zedruu / Xyris / Isshin / The Worst Possible Commander Show 23 Magic MTG",
                },
            ),
            content: [
                MediaContent {
                    url: Some(
                        Url {
                            scheme: "https",
                            cannot_be_a_base: false,
                            username: "",
                            password: None,
                            host: Some(
                                Domain(
                                    "www.youtube.com",
                                ),
                            ),
                            port: None,
                            path: "/v/EwEUvsOnKBI",
                            query: Some(
                                "version=3",
                            ),
                            fragment: None,
                        },
                    ),
                    content_type: Some(
                        "application/x-shockwave-flash",
                    ),
                    height: Some(
                        390,
                    ),
                    width: Some(
                        640,
                    ),
                    duration: None,
                    size: None,
                    rating: None,
                },
            ],
            duration: None,
            thumbnails: [
                MediaThumbnail {
                    image: Image {
                        uri: "https://i2.ytimg.com/vi/EwEUvsOnKBI/hqdefault.jpg",
                        title: None,
                        link: None,
                        width: Some(
                            480,
                        ),
                        height: Some(
                            360,
                        ),
                        description: None,
                    },
                    time: None,
                },
            ],
            texts: [],
            description: Some(
                Text {
                    content_type: "text/plain",
                    src: None,
                    content: "SUPPORT** Use Promo Code WORST for 5% off at CoolStuffInc.com\nhttps://www.coolstuffinc.com/\nPatreon - https://www.patreon.com/covertgocrew\nDecklists: \nCGB - Laelia, Blade Reforged Mono Red Aggro\nhttps://www.moxfield.com/decks/3M1HZnzjG0ShsuYLLt-_eA\nBlake - Zedruu the Greathearted Jeskai Gifts Control\nhttps://www.moxfield.com/decks/xnyJ38VmLku2FJ7T3P5nog\nShea -  Xyris, the Writhing Storm Temur Snakes Tribal\nhttps://www.moxfield.com/decks/-ew6hG2JMUaUT64zUIUDAA\nBen - Isshin, Two Heavens as One Eldrazi Annihilator\nhttps://www.moxfield.com/decks/QHkqgZQjrEW1If3RrMUCkw\nTwitters:\nCGB: @covertgoblue\nBlake: @blakebandro\nBen: @GreedLordMTG\nShea: @TheWorstShea\n\nTitle Music by Sam Plotkin - www.samplotkinmusic.com\nAdditional music by Rattlesnake\nLogo design by HolyMTG - HolyMTG@gmail.com Twitter @HolyMTG\n\nUnofficial Fan Content permitted under the Fan Content Policy. Not approved/endorsed by Wizards. Portions of the materials used are property of Wizards of the Coast. ©Wizards of the Coast LLC.\n\n#cgb #mtgarena #mtg",
                },
            ),
            community: Some(
                MediaCommunity {
                    stars_avg: Some(
                        5.0,
                    ),
                    stars_count: Some(
                        1720,
                    ),
                    stars_min: Some(
                        1,
                    ),
                    stars_max: Some(
                        5,
                    ),
                    stats_views: Some(
                        25202,
                    ),
                    stats_favorites: None,
                },
            ),
            credits: [],
        },
    ],
},
```

That's why I assumed FreshRSS must have the same information and could include the thumbnail as an enclosure.

**Comment by Alkarex** [comment]
I have not understood the question :-)
What I wrote higher up is that when a feed provide enclosures, there are supposed to be in our Google Reader API responses (except images).

**Comment by jangernert** [comment]
Okay, sorry. I guess my question is answered by looking at the code again and using my brain for a change.

So yes, it would indeed be nice to also include image-enclosures in the greader API :)

I have very little knowledge about PHP and web development in general. But I do have a question about the code if you don't mind:  It looks to me like enclosures are transformed into a few HTML tags and then merged with the actual content of the article. And when that same article is later on queried via API: the HTML is parsed with libxml2 and the enclosures are extracted via XPath query. Isn't that a lot of extra work compared to storing enclosures in separate table in the database?

**Comment by Alkarex** [comment]
Indeed, reworking enclosures is on my todo list. I have recently added a new attribute to our database entries https://github.com/FreshRSS/FreshRSS/pull/4444 which is appropriate for storing lists of enclosures.

**Comment by jangernert** [comment]
Cool! Looking forward to it.

In the mean time I'll implement enclosures for FreshRSS in NewsFlash. Since it is not part of the original google reader API (at least I didn't find anything about it) I'd like to contribute a few lines of documentation about enclosures to `06_GoogleReader_API.md` once I am done with the implementation.

**Comment by Alkarex** [comment]
Enclosures were a part of the original Google Reader API. See e.g.:
* http://web.archive.org/web/20100327173442/http://blog.martindoms.com/2009/10/16/using-the-google-reader-api-part-2/
* https://github.com/noinnion/newsplus/blob/2ce0002c8f4cea594fe1208922e1f4f184e98eb2/extensions/GoogleReaderCloneExtension/src/com/noinnion/android/newsplus/extension/google_reader/GoogleReaderClient.java#L760-L778


Here is an example from my old Google Reader archive:

```json
{
	"isReadStateLocked": true,
	"crawlTimeMsec": "1308572644075",
	"timestampUsec": "1308572644075475",
	"id": "tag:google.com,2005:reader/item/2f3f9b887c598f5a",
	"categories": ["user/10311923250980613279/state/com.google/broadcast", "user/10311923250980613279/state/com.google/like", "user/10311923250980613279/state/com.google/read", "Article - Computing"],
	"title": "Ultrasound App Lets Almost Any Phone Pay",
	"published": 1308567600,
	"updated": 1308567600,
	"enclosure": [{
			"href": "http://www.pheedo.com/e/406947017b9863a34dd04196feaf22bd/37860.mp3",
			"type": "audio/mpeg",
			"length": "2500000"
		}, {
			"href": "http://www.pheedo.com/e/b76ab942762f5d56f2e49eee8e5e84ed/zpay_x116.jpg",
			"type": "image/jpeg",
			"length": "2750"
		}
	],
	"alternate": [{
			"href": "http://feeds.technologyreview.com/click.phdo?i=fa6f300ed23fd6a17f49194b43afd652",
			"type": "text/html"
		}
	],
	"summary": {
		"direction": "ltr",
		"content": "<p>Only a few handsets contain contactless payment chips, but many more devices could use sounds to achieve the same purpose.</p>\n<p>For years, banks, cell-phone carriers, and tech companies have been experimenting with technology that enables mobile phones to <a href=\"http://www.technologyreview.com/communications/26493/\">connect securely over short distances and make payments in stores</a>. Although <a href=\"http://www.technologyreview.com/communications/37636/\">Google and other big companies</a> say they're committed to the radio-based technology known as near-field communications (NFC), few phones have the necessary hardware built in.</p><br style=\"clear:both\">\n<br style=\"clear:both\">\n<a href=\"http://ads.pheedo.com/click.phdo?s=fa6f300ed23fd6a17f49194b43afd652&amp;p=1\"><img alt=\"\" style=\"border:0\" border=\"0\" src=\"http://ads.pheedo.com/img.phdo?s=fa6f300ed23fd6a17f49194b43afd652&amp;p=1\"></a>\n<img alt=\"\" height=\"0\" width=\"0\" border=\"0\" src=\"http://segment-pixel.invitemedia.com/pixel?code=TechBiz&amp;partnerID=167&amp;key=segment\"><img alt=\"\" height=\"0\" width=\"0\" border=\"0\" src=\"http://pixel.quantserve.com/pixel/p-8bUhLiluj0fAw.gif?labels=pub.28834.rss.TechBiz.7020,cat.TechBiz.rss\">"
	},
	"comments": [],
	"annotations": [],
	"origin": {
		"streamId": "feed/http://www.technologyreview.com/rss/rss.aspx",
		"title": "New on MIT Technology Review",
		"htmlUrl": "http://www.technologyreview.com/stream/?sort=recent"
	}
}
```

P.S. The article of the example is still online :-) https://www.technologyreview.com/2011/06/20/118435/ultrasound-app-lets-almost-any-phone-pay/

**Comment by jangernert** [comment]
I implemented support for enclosures on the side of NewsFlash and noticed that sometimes the enclosure has a name `<p class="enclosure-title">`. Would be nice to extend the enclosure object with the `title`. Otherwise every enclosure would be nameless like this:

![Screenshot from 2022-10-23 17-09-54](https://user-images.githubusercontent.com/12012494/197400148-2455819a-76ca-437c-ab39-08ed413b57d0.png)



Maybe that's already the case. All of the podcasts that I found didn't have a title for enclosures. And all the other enclosures weren't showing up in the json output of FreshRSS.
